### PR TITLE
Buttons for authorization request webhook messages

### PR DESCRIPTION
**Pull request by danielpeng2**
## Description

This change adds interactive buttons to the message for dataset authorization request webhooks. These buttons allow users to accept/reject/cancel these dataset authorization requests through the data.world Public API, as long as they are properly authorized.

Scenario | View
-|-
User requests access to dataset | Screen Shot 2020-12-22 at 3 40 44 PM
Admin invites user to dataset | Screen Shot 2020-12-22 at 3 43 06 PM
Admin acts on request | Screen Shot 2020-12-22 at 3 43 52 PM
Unauthorized user acts on request | Screen Shot 2020-12-22 at 3 45 25 PM
Generic error message (e.g. request has already been acted through data.world website) | Screen Shot 2020-12-22 at 3 46 47 PM

## Testing

1. Run the Slack app in a workspace
2. Invite the Slack app to a channel (`/invite @dataworld`)
3. Associate your Slack account with a data.world account (running `/data.world webhook` will prompt this)
4. In the Slack channel, run `/data.world webhook`.
5. In an org that your associated data.world user is an admin of:
  a. Add the webhook URL as an org-level webhook
  b. Invite a user to access a dataset. You should receive a message in the Slack channel - cancel this request
  c. As a different user, request access to a dataset. You should receive a message in the Slack channel - approve/reject the request.
6. In an org that your associated data.world user is not an admin of:
  a. Add the webhook URL as an org-level webhook
  b. Invite a user to access a dataset. You should receive a message in the Slack channel - try cancelling this request. A modal should appear saying you are unauthorized.

### CODE REVIEW

File: server/controllers/command.js [hunk 1]
```diff
@@ -722,39 +724,70 @@ const performAction = async (req, res) => {
...
+      collection.forEach(payload.actions, async action => {
+        const actionid = action.action_id
+        if (Object.values(AUTHORIZATION_ACTIONS).includes(actionid)) {
+          const { requestid, agentid, datasetid } = JSON.parse(action.value);
+          await handleDatasetRequestAction({
+            channelid: payload.channel.id,
+            userid: payload.user.id,
+            triggerid: payload.trigger_id,
+            responseUrl: payload.response_url,
+            message: payload.message,
```
Comment (danielpeng2) [inline server/controllers/command.js:779]: @katz12 I overlooked that the original message is actually passed in the action payload we receive. So we don't have to add the extra `history` scopes to update the original message.

In order to fully support threading later on, we may need to add the `history` scopes.

Comment (katz12) [reply server/controllers/command.js:779]: Oh nice!

Why would we need the scope for threading then? Or are you just unsure whether we have everything we need in this message?

File: server/api/slack.js [hunk 2]
```diff
@@ -267,15 +267,20 @@ const sendMessageWithAttachments = (botAccessToken, channelId, attachments) => {
...
-const sendMessageWithBlocks = (botAccessToken, channelId, blocks) => {
+const sendMessageWithBlocks = async (botAccessToken, channelId, blocks) => {
   const slackBot = new SlackWebClient(botAccessToken);
-  slackBot.chat.postMessage({
+  await slackBot.chat.postMessage({
     channel: channelId,
     text: "",
     blocks
   });
 }
```
Comment (katz12) [inline server/api/slack.js:277]: Why did you need to make this async? You don't seem to be waiting on it.

File: server/controllers/command.js (hunk 1 above)
Comment (danielpeng2) [reply server/controllers/command.js:779]: Not specifically for threading, per se. It would be useful for handling multiple notifications about the same request.

Let's say a user requests access to a dataset and the request is accepted on data.world (not through Slack). Slack would receive an `authorization_request.created` message followed by an `authorization_request.accepted` message some time later. Since the request was not accepted on Slack, the original `authorization_request.created` Slack message would not update. Users could still try to accept or approve the request on Slack, but this would show error messages, as the request is now invalid.

Ideally, we would update the original message `authorization_request.created` after we receive an `authorization_request.accepted` message for the same request. But we don't have access to the `authorization_request.created` message this time, so we would need to store the request data to build the original message again, or use the `history` scope to fetch the original message.

File: server/controllers/commands/request.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      await openNotificationModal(
+        channelid,
+        triggerid,
+        'You are not authorized to manage this request.'
+      )
+    } else {
+      await openNotificationModal(
+        channelid,
+        triggerid,
+        'Could not successfully manage this request. Try viewing the request on <https://data.world|data.world>.'
```
Comment (katz12) [inline server/controllers/commands/request.js:82]: nit: this won't be an issue right now, but this hard-coded url will need to change eventually.

File: server/api/slack.js (hunk 2 above)
Comment (danielpeng2) [reply server/api/slack.js:277]: I think my intention was to wait on `handleAuthorizationRequest` and `handleContributionRequest`, so that errors would be caught properly. I must've forgot to propagate the `await`s all the way up.

https://github.com/datadotworld/slack-app/blob/9bf040c4af585fd598777ee95dbdbaff8f50004e/server/controllers/webhook.js#L799-L813

File: server/controllers/commands/request.js (hunk 3 above)
Comment (danielpeng2) [reply server/controllers/commands/request.js:82]: A lot of the pre-existing features rely on hard-coded `data.world` URLs (e.g. link previews and dataset/project subscriptions). It might be good to handle these all at once, once we know more about how the sites functionality will be implemented.

File: server/jest/fixtures.js [hunk 4]
```diff
@@ -45,6 +45,31 @@ const authorizationRequestCancelledEventBody = {
...
+  resourceId: "use-case-discvoerable",
+  resourceName: "use-case-discvoerable",
+  resourceOwner: "use-case-org",
+  resourceOwnerName: "Use Case Org",
+  resourceUrl: "http://localhost:3000/use-case-org/use-case-discvoerable",
+  event: {
+    triggeredBy: "user8888",
+    created: "Sat, 12 Dec 2020 21:36:48 GMT"
+  },
+  eventType: "dataset.authorization_invite.created",
```
Comment (katz12) [inline server/jest/fixtures.js:64]: Am I misunderstanding what this fixture is for? Should the event type be "created" for a cancelled event?

**Review by katz12 — COMMENTED** [review]
This all seems pretty solid to me. @kennyshittu do you have any thoughts?

File: server/controllers/commands/request.js (hunk 3 above)
Comment (katz12) [reply server/controllers/commands/request.js:82]: Yep agreed.

File: server/jest/fixtures.js (hunk 4 above)
Comment (danielpeng2) [reply server/jest/fixtures.js:64]: Good catch, the event type should be "cancelled" here.

File: server/controllers/command.js (hunk 1 above)
Comment (katz12) [reply server/controllers/command.js:779]: > Let's say a user requests access to a dataset and the request is accepted on data.world (not through Slack).

Yeah I was thinking about this case. This makes sense for future functionality, and I see that it is different than a message being acted upon because there is no reference to the slack message in the raw d.w webhook.

File: server/jest/fixtures.js (hunk 4 above)
Comment (danielpeng2) [reply server/jest/fixtures.js:64]: Looking at it again, I think the content of the fixture was right, I just misnamed it. I wanted to test that the `cancel` button was showing correctly on authorization invites.

Comment (katz12) [reply server/jest/fixtures.js:64]: That makes more sense. The event type here should be driving the functionality.

File: server/controllers/command.js (hunk 1 above)
Comment (danielpeng2) [reply server/controllers/command.js:779]: Yep, this is a scenario where we would likely need to create a requests table to store the id of the parent message.

**Review by kennyshittu — APPROVED** [review]
